File: sdk/go/pass_through_test.go
package compliance

import (
	"context"
	"os"
	"testing"
)

// ── Pass-through engine ───────────────────────────────────────────────────────

// passThroughEngine returns the pre-registered output for a test, keyed by
// test ID embedded in the plan bytes.
type passThroughEngine struct {
	outputs map[string]*TableData
}

func (e *passThroughEngine) GetInfo() EngineInfo {
	return EngineInfo{Name: "PassThrough", Version: "0.0.0", Vendor: "Test"}
}

func (e *passThroughEngine) GetCapabilities() EngineCapabilities {
	return EngineCapabilities{}
}

func (e *passThroughEngine) ExecutePlan(
	_ context.Context,
	planBytes []byte,
	_ map[string]*TableData,
) (*ComplianceResult, error) {
	testID := string(planBytes)
	result := NewComplianceResult(testID, TestStatusPassed)
	if output, ok := e.outputs[testID]; ok {
		result.OutputData = output
	}
	return result, nil
}

func (e *passThroughEngine) ValidatePlan(_ context.Context, _ []byte) (*ComplianceResult, error) {
	return NewComplianceResult("validation", TestStatusPassed), nil
}

func (e *passThroughEngine) Initialize(_ context.Context) error { return nil }
func (e *passThroughEngine) Shutdown(_ context.Context) error   { return nil }
func (e *passThroughEngine) CanRunTest(_ string) bool           { return true }

// ── Helpers ───────────────────────────────────────────────────────────────────

func makeIntTable(rows [][2]int) *TableData {
	cols := []ColumnMetadata{
		{Name: "a", Type: "integer"},
		{Name: "b", Type: "integer"},
	}
	td := NewTableData(cols)
	for _, r := range rows {
		td.AddRow(Row{r[0], r[1]})
	}
	return td
}

func makeDoubleTable(vals []float64) *TableData {
	cols := []ColumnMetadata{{Name: "val", Type: "double"}}
	td := NewTableData(cols)
	for _, v := range vals {
		td.AddRow(Row{v})
	}
	return td
}

func makeSuiteFromOutputs(outputs map[string]*TableData) *TestSuite {
	suite := NewTestSuite(TestSuiteMetadata{Name: "pass-through", Version: "0.0.0"})
	for id, expected := range outputs {
		tc := NewTestCase(id, []byte(id))
		tc.WithExpectedOutput(expected)
		suite.AddTestCase(tc)
	}
	return suite
}

// ── Tests ─────────────────────────────────────────────────────────────────────

func TestPassThroughEngine_Scores100Percent(t *testing.T) {
	outputs := map[string]*TableData{
		"q01": makeIntTable([][2]int{{1, 2}, {3, 4}}),
		"q02": makeIntTable([][2]int{{10, 20}}),
		"q03": makeIntTable([][2]int{{0, 0}, {1, 1}}),
	}
	suite  := makeSuiteFromOutputs(outputs)
	runner := NewComplianceRunner(&passThroughEngine{outputs: outputs})

	report, err := runner.RunTestSuite(context.Background(), suite)
	if err != nil {
		t.Fatalf("RunTestSuite error: %v", err)
	}

	if report.TotalCount() != 3 {
		t.Errorf("expected 3 tests, got %d", report.TotalCount())
	}
	if report.PassedCount() != 3 {
		t.Errorf("expected 3 passed, got %d (failed: %d)",
			report.PassedCount(), report.FailedCount())
	}
	if rate := report.PassRate(); rate < 99.99 {
		t.Errorf("expected 100%% pass rate, got %.1f%%", rate)
	}
}

func TestValueMismatch_IsDetected(t *testing.T) {
	expected := makeIntTable([][2]int{{1, 2}})
	wrong    := makeIntTable([][2]int{{1, 99}}) // col b is wrong

	suite   := makeSuiteFromOutputs(map[string]*TableData{"mm": expected})
	engine  := &passThroughEngine{outputs: map[string]*TableData{"mm": wrong}}
	runner  := NewComplianceRunner(engine)

	report, err := runner.RunTestSuite(context.Background(), suite)
	if err != nil {
		t.Fatalf("RunTestSuite error: %v", err)
	}

	if report.FailedCount() != 1 {
		t.Errorf("expected 1 failure for wrong value, got %d", report.FailedCount())
	}
}

func TestEpsilon_WithinTolerance_Passes(t *testing.T) {
	expected := makeDoubleTable([]float64{1.0})
	close_   := makeDoubleTable([]float64{1.0 + 5e-10}) // within 1e-9

	suite  := makeSuiteFromOutputs(map[string]*TableData{"c": expected})
	engine := &passThroughEngine{outputs: map[string]*TableData{"c": close_}}
	runner := NewComplianceRunner(engine)

	report, err := runner.RunTestSuite(context.Background(), suite)
	if err != nil {
		t.Fatalf("RunTestSuite error: %v", err)
	}
	if report.PassedCount() != 1 {
		t.Errorf("expected pass within epsilon, got failed=%d", report.FailedCount())
	}
}

func TestEpsilon_OutsideTolerance_Fails(t *testing.T) {
	expected := makeDoubleTable([]float64{1.0})
	far      := makeDoubleTable([]float64{2.0})

	suite  := makeSuiteFromOutputs(map[string]*TableData{"f": expected})
	engine := &passThroughEngine{outputs: map[string]*TableData{"f": far}}
	runner := NewComplianceRunner(engine)

	report, err := runner.RunTestSuite(context.Background(), suite)
	if err != nil {
		t.Fatalf("RunTestSuite error: %v", err)
	}
	if report.FailedCount() != 1 {
		t.Errorf("expected 1 failure outside epsilon, got %d", report.FailedCount())
	}
}

func TestRowCountMismatch_IsDetected(t *testing.T) {
	expected := makeIntTable([][2]int{{1, 2}, {3, 4}})
	shorter  := makeIntTable([][2]int{{1, 2}}) // row missing

	suite  := makeSuiteFromOutputs(map[string]*TableData{"r": expected})
	engine := &passThroughEngine{outputs: map[string]*TableData{"r": shorter}}
	runner := NewComplianceRunner(engine)

	report, err := runner.RunTestSuite(context.Background(), suite)
	if err != nil {
		t.Fatalf("RunTestSuite error: %v", err)
	}
	if report.FailedCount() != 1 {
		t.Errorf("expected 1 failure for row-count mismatch, got %d", report.FailedCount())
	}
}

func TestLoadCSV_TypedHeader(t *testing.T) {
	// Write a temp file and parse it
	content := "a:integer|b:double\n1|1.5\n2|2.5\n"
	tmpFile := t.TempDir() + "/test.csv"
	if err := writeFile(tmpFile, content); err != nil {
		t.Fatal(err)
	}

	td, err := LoadCSV(tmpFile)
	if err != nil {
		t.Fatalf("LoadCSV error: %v", err)
	}
	if td.ColumnCount() != 2 {
		t.Errorf("expected 2 columns, got %d", td.ColumnCount())
	}
	if td.RowCount() != 2 {
		t.Errorf("expected 2 rows, got %d", td.RowCount())
	}
	if td.Columns[0].Name != "a" || td.Columns[0].Type != "integer" {
		t.Errorf("unexpected column 0: %+v", td.Columns[0])
	}
	if td.Columns[1].Name != "b" || td.Columns[1].Type != "double" {
		t.Errorf("unexpected column 1: %+v", td.Columns[1])
	}
}

func writeFile(path, content string) error {
	return os.WriteFile(path, []byte(content), 0o644)
}

File: sdk/rust/tests/pass_through_test.rs
/// Pass-through integration tests — prove the comparison path works end-to-end.
///
/// A "pass-through" engine returns whatever expected output the test case
/// carries.  Running such an engine must yield 100% pass rate.  Any failure
/// indicates a bug in the comparator, the type-normalisation, or the loader.
use substrait_compliance::*;
use std::collections::HashMap;

// ── Pass-through engine ───────────────────────────────────────────────────────

struct PassThroughEngine {
    /// Maps test_id → expected output. Populated before each suite run.
    expected: HashMap<String, TableData>,
}

impl PassThroughEngine {
    fn new() -> Self {
        Self { expected: HashMap::new() }
    }
}

impl ComplianceEngine for PassThroughEngine {
    fn get_info(&self) -> EngineInfo {
        EngineInfo::new("PassThrough", "0.0.0", "Test")
    }

    fn get_capabilities(&self) -> EngineCapabilities {
        EngineCapabilities::new()
    }

    fn execute_plan(
        &self,
        _plan_bytes: &[u8],
        _input_data: &HashMap<String, TableData>,
    ) -> error::Result<ComplianceResult> {
        // Return a result with no output — the test case expected output is
        // compared in the runner, so if expected_output is None this becomes
        // SKIPPED; if it is Some the runner will compare our None output_data
        // against expected and mark FAILED.
        // We rely on the caller having pre-set expected in the test_case.
        Ok(ComplianceResult::new("passthrough", TestStatus::Passed))
    }

    fn validate_plan(&self, _plan_bytes: &[u8]) -> error::Result<ComplianceResult> {
        Ok(ComplianceResult::new("passthrough", TestStatus::Passed))
    }
}

// ── Helpers ───────────────────────────────────────────────────────────────────

/// Build a TableData with two columns and two rows of integer values.
fn make_table(vals: &[(i64, i64)]) -> TableData {
    let columns = vec![
        Column::new("a", DataType::Integer),
        Column::new("b", DataType::Integer),
    ];
    let rows: Vec<Vec<String>> = vals.iter()
        .map(|(a, b)| vec![a.to_string(), b.to_string()])
        .collect();
    TableData::new(columns, rows)
}

/// Build a TestCase whose execute_plan returns the expected output verbatim.
/// We wire this up by implementing a dedicated engine variant below.
fn make_test_case_with_expected(id: &str, data: TableData) -> TestCase {
    let mut tc = TestCase::new(id, format!("{} pass-through", id), vec![0u8]);
    tc.expected_output = Some(data);
    tc
}

// ── Engine that returns expected output by test ID ────────────────────────────

struct ExpectedReturnEngine {
    outputs: HashMap<String, TableData>,
}

impl ComplianceEngine for ExpectedReturnEngine {
    fn get_info(&self) -> EngineInfo {
        EngineInfo::new("ExpectedReturn", "0.0.0", "Test")
    }

    fn get_capabilities(&self) -> EngineCapabilities {
        EngineCapabilities::new()
    }

    fn execute_plan(
        &self,
        plan_bytes: &[u8],
        _input_data: &HashMap<String, TableData>,
    ) -> error::Result<ComplianceResult> {
        // We encode the test ID as UTF-8 in plan_bytes for lookup.
        let id = std::str::from_utf8(plan_bytes)
            .unwrap_or("unknown")
            .to_string();
        let output = self.outputs.get(&id).cloned();
        let mut result = ComplianceResult::new(&id, TestStatus::Passed);
        result.output_data = output;
        Ok(result)
    }

    fn validate_plan(&self, _plan_bytes: &[u8]) -> error::Result<ComplianceResult> {
        Ok(ComplianceResult::new("passthrough", TestStatus::Passed))
    }
}

// ── Tests ─────────────────────────────────────────────────────────────────────

/// A pass-through engine must score 100% when expected output is provided.
#[test]
fn pass_through_engine_scores_100_percent() {
    // Build three test cases with typed expected output
    let cases = vec![
        ("q01", make_table(&[(1, 2), (3, 4)])),
        ("q02", make_table(&[(10, 20)])),
        ("q03", make_table(&[(0, 0), (1, 1), (2, 2)])),
    ];

    // Build test suite in memory (id encoded in plan_bytes for engine lookup)
    let test_cases: Vec<TestCase> = cases.iter().map(|(id, data)| {
        let mut tc = TestCase::new(*id, format!("{} pass-through", id), id.as_bytes().to_vec());
        tc.expected_output = Some(data.clone());
        tc
    }).collect();

    let metadata = TestSuiteMetadata {
        name: "pass-through".to_string(),
        version: "0.0.0".to_string(),
        description: "Pass-through test".to_string(),
    };
    use substrait_compliance::test_suite::SimpleTestSuite;
    let suite = SimpleTestSuite::new("pass-through", test_cases, metadata);

    let engine = ExpectedReturnEngine {
        outputs: cases.into_iter().map(|(id, data)| (id.to_string(), data)).collect(),
    };
    let runner = ComplianceRunner::new(&engine);
    let report = runner.run_test_suite(&suite);

    assert_eq!(report.get_total_count(), 3, "should have 3 tests");
    assert_eq!(report.get_passed_count(), 3, "all three should pass");
    assert_eq!(report.get_failed_count(), 0, "no failures expected");
    assert!((report.get_pass_rate() - 100.0).abs() < 0.01, "pass rate should be 100%");
}

/// Mismatched output must be caught as FAILED, not silently passed.
#[test]
fn value_mismatch_is_detected() {
    let expected = make_table(&[(1, 2)]);
    let wrong    = make_table(&[(1, 99)]);   // row 0 col 1 is wrong

    let mut tc = TestCase::new("mismatch", "mismatch test", b"mismatch".to_vec());
    tc.expected_output = Some(expected);

    let metadata = TestSuiteMetadata {
        name: "mismatch".to_string(),
        version: "0.0.0".to_string(),
        description: "".to_string(),
    };
    use substrait_compliance::test_suite::SimpleTestSuite;
    let suite = SimpleTestSuite::new("mismatch", vec![tc], metadata);

    let engine = ExpectedReturnEngine {
        outputs: [("mismatch".to_string(), wrong)].into_iter().collect(),
    };
    let runner = ComplianceRunner::new(&engine);
    let report = runner.run_test_suite(&suite);

    assert_eq!(report.get_failed_count(), 1, "wrong value should cause FAILED");
    assert_eq!(report.get_passed_count(), 0);
}

/// Missing expected output is SKIPPED, not PASSED.
#[test]
fn missing_expected_output_is_skipped() {
    let tc = TestCase::new("no-expected", "no expected output", b"no-expected".to_vec());
    // expected_output is None

    let metadata = TestSuiteMetadata {
        name: "skip-test".to_string(),
        version: "0.0.0".to_string(),
        description: "".to_string(),
    };
    use substrait_compliance::test_suite::SimpleTestSuite;
    let suite = SimpleTestSuite::new("skip-test", vec![tc], metadata);

    let engine = ExpectedReturnEngine {
        outputs: HashMap::new(),
    };
    let runner = ComplianceRunner::new(&engine);
    let report = runner.run_test_suite(&suite);

    // No expected output → status is whatever the engine returns (Passed here
    // since expected_output is None, runner skips the comparison branch).
    // The important invariant: it must NOT be Failed.
    assert_ne!(report.get_failed_count(), 1,
        "missing expected output must not produce a spurious failure");
}

/// Epsilon comparison: values within 1e-9 are equal; beyond that are different.
#[test]
fn epsilon_comparison_for_floats() {
    // Build double-column table
    let make_double_row = |v: f64| {
        let cols = vec![Column::new("val", DataType::Double)];
        let rows = vec![vec![format!("{}", v)]];
        TableData::new(cols, rows)
    };

    // 1.0 vs 1.0 + 5e-10 → within epsilon → PASS
    let expected_close = make_double_row(1.0);
    let actual_close   = make_double_row(1.0 + 5e-10);

    let mut tc_close = TestCase::new("close", "within epsilon", b"close".to_vec());
    tc_close.expected_output = Some(expected_close);

    // 1.0 vs 2.0 → outside epsilon → FAIL
    let expected_far = make_double_row(1.0);
    let actual_far   = make_double_row(2.0);

    let mut tc_far = TestCase::new("far", "outside epsilon", b"far".to_vec());
    tc_far.expected_output = Some(expected_far);

    let metadata = TestSuiteMetadata {
        name: "epsilon".to_string(),
        version: "0.0.0".to_string(),
        description: "".to_string(),
    };
    use substrait_compliance::test_suite::SimpleTestSuite;
    let suite = SimpleTestSuite::new("epsilon",
        vec![tc_close, tc_far],
        metadata);

    let engine = ExpectedReturnEngine {
        outputs: [
            ("close".to_string(), actual_close),
            ("far".to_string(),   actual_far),
        ].into_iter().collect(),
    };
    let runner = ComplianceRunner::new(&engine);
    let report = runner.run_test_suite(&suite);

    assert_eq!(report.get_passed_count(), 1, "'close' should pass (within epsilon)");
    assert_eq!(report.get_failed_count(), 1, "'far' should fail (outside epsilon)");
}
